tensorflow_addons/layers/optical_flow_test.py

```diff
@@ -0,0 +1,199 @@
+            input_b = tf.constant(valb, dtype=tf.float32)
+
+            if data_format == 'channels_last':
+                input_a = tf.transpose(input_a, [0, 2, 3, 1])
+                input_b = tf.transpose(input_b, [0, 2, 3, 1])
+
+            input_a_tensor = tf.convert_to_tensor(input_a, dtype=tf.float32)
+            input_b_tensor = tf.convert_to_tensor(input_b, dtype=tf.float32)
+
+            kernel_size = 1
+            max_displacement = 2
+            stride_1 = 1
+            stride_2 = 2
+            pad = 4
+
+            actual = self._forward(
+                input_a_tensor,
+                input_b_tensor,
+                kernel_size=kernel_size,
+                max_displacement=max_displacement,
+                stride_1=stride_1,
+                stride_2=stride_2,
+                pad=pad,
+                data_format=data_format)
+
+            if data_format == 'channels_last':
+                # NHWC -> NCHW
+                actual = tf.transpose(actual, [0, 3, 1, 2])
+
+            # We can test fixed ids, as output is independent from data_format
+            expected_ids = np.concatenate([np.zeros(464,), np.ones(464,)])
+            self.assertAllClose(
+                tf.where(tf.equal(actual, 0))[:, 0], expected_ids)
+
+            counts = [54, 52, 54, 50, 44, 50, 54, 52, 54]
+            expected_ids = np.concatenate(
+                [k * np.ones(v,) for k, v in enumerate(counts)])
+            expected_ids = np.concatenate([expected_ids, expected_ids])
+            self.assertAllClose(
+                tf.where(tf.equal(actual, 0))[:, 1], expected_ids)
+            self.assertEqual(actual.shape, (2, 9, 7, 8))
+
+    def _gradients(self, data_format):
+        with test_utils.use_gpu():
+            batch, channels, height, width = 2, 3, 5, 6
+            input_a = np.random.randn(batch, channels, height,
+                                      width).astype(np.float32)
+            input_b = np.random.randn(batch, channels, height,
+                                      width).astype(np.float32)
+
+            kernel_size = 1
+            max_displacement = 2
+            stride_1 = 1
+            stride_2 = 2
+            pad = 4
+
+            if data_format == 'channels_last':
+                input_a = tf.transpose(input_a, [0, 2, 3, 1])
+                input_b = tf.transpose(input_b, [0, 2, 3, 1])
+
+            input_a_op = tf.convert_to_tensor(input_a)
+            input_b_op = tf.convert_to_tensor(input_b)
+
+            def correlation_fn(input_a, input_b):
+                return correlation_cost(
+                    input_a,
+                    input_b,
+                    kernel_size=kernel_size,
+                    max_displacement=max_displacement,
+                    stride_1=stride_1,
+                    stride_2=stride_2,
+                    pad=pad,
+                    data_format=data_format)
+
+            theoretical, numerical = tf.test.compute_gradient(
```

Comment:
I would rename `theoretical` to `gradient_from_implementation` and `numerical` to `numerical_gradient`.

Reply:
I do like that re-name, but going to leave it so its consistent with TF Core:
https://github.com/tensorflow/tensorflow/search?utf8=%E2%9C%93&q=theoretical%2C+numerical&type=